flink-dstl/flink-dstl-dfs/src/test/java/org/apache/flink/changelog/fs/FsStateChangelogWriterTest.java

```diff
@@ -201,6 +201,29 @@ void testPersistNonFailedChanges() throws Exception {
                 });
     }
 
+    @Test
+    void testPreEmptivelyUpload() throws Exception {
+        int preEmptivelyUploadThreshold = 100;
+        withWriter(
+                preEmptivelyUploadThreshold,
+                (writer, uploader) -> {
+                    writer.append(KEY_GROUP, getBytes(10));
+
+                    // trigger materialization, move ChangeSet[sqn=0] to notUploaded
+                    SequenceNumber materializationSqn = writer.nextSequenceNumber();
+
+                    // trigger pre-emptively upload
+                    writer.append(KEY_GROUP, getBytes(100));
+
+                    // assert pre-emptively upload happened
+                    SequenceNumber sqnAfterPreEmptivelyUpload = writer.lastAppendedSqnUnsafe();
+                    assertThat(sqnAfterPreEmptivelyUpload).isEqualTo(materializationSqn.next());
+
+                    // assert pre-emptively upload not persisted older data
+                    assertThat(writer.getNotUploaded()).isNotEmpty();
```

Comment:
Could you please add some checks on when the notUploaded StateChangeSet will be deleted?
Or move https://github.com/apache/flink/pull/21812/files#diff-cc303bbf16bfae7907b8983d0ef8ab52b964135d5e3efec60515d0cf377fab66R113 to here?